File: swift/Sources/CircleAI/WindowsAutomation.swift
// WindowsAutomation.swift
//
// Port of src/CircleAI.WindowsAutomation/:
//   • Contracts.cs                 → UiElement, IUiAutomationDriver
//   • InMemoryWindowsAutomation.cs → UiAutomationEvent, InMemoryUiAutomationDriver
//   • NullImplementations.cs       → NullUiAutomationDriver
//   • WindowsAutomationHelpers.cs  → UiElementHelpers (hit-test / dump)
//
// The package is OS-neutral in shape: hosts snap a real Win32-UIA driver in for
// production; the in-memory driver lets callers drive a virtual UI without a
// desktop. Nothing here is Windows-specific at the Swift level — it is the
// contract + a portable reference driver, so it belongs in the portable surface.
//
// Porting notes:
//   • C# `ValueTask` / `ValueTask<T>` → `async` / `async -> T`.
//   • `IObserver`-style `Observe(Action<…>)` → an `@escaping @Sendable` closure.
//     Observer notifications swallow thrown errors, matching the C# try/catch
//     that logs and continues.
//   • `InMemoryUiAutomationDriver` holds mutable element + observer state →
//     `final class … @unchecked Sendable` + a single `NSLock`. The snapshot
//     copies observers under the lock before invoking them (as the C# does).
//   • Argument validation (`elementId required`, unknown-element, non-null text)
//     is reproduced via `precondition` / thrown `UiAutomationError`.

import Foundation

// MARK: - UiElement

/// A single UI-automation element. Port of C# record `UiElement`.
public struct UiElement: Sendable, Equatable {
    public let elementId: String
    public let name: String
    public let kind: String
    public let x: Int
    public let y: Int
    public let width: Int
    public let height: Int

    public init(elementId: String, name: String, kind: String, x: Int, y: Int, width: Int, height: Int) {
        self.elementId = elementId
        self.name = name
        self.kind = kind
        self.x = x
        self.y = y
        self.width = width
        self.height = height
    }
}

// MARK: - IUiAutomationDriver

/// Cross-backend UI-automation driver contract. Port of C# `IUiAutomationDriver`.
public protocol IUiAutomationDriver: AnyObject, Sendable {
    var backendId: String { get }
    func snapshot() async throws -> [UiElement]
    func click(_ elementId: String) async throws
    func type(_ text: String) async throws
    func key(_ keyName: String) async throws
}

// MARK: - Errors

public enum UiAutomationError: Error, Equatable, CustomStringConvertible {
    case elementIdRequired
    case unknownElement(String)
    case keyNameRequired

    public var description: String {
        switch self {
        case .elementIdRequired: return "elementId required"
        case .unknownElement(let id): return "Unknown element '\(id)'."
        case .keyNameRequired: return "keyName required"
        }
    }
}

// MARK: - UiAutomationEvent

/// An observable event raised by the in-memory driver. Port of C# record
/// `UiAutomationEvent`.
public struct UiAutomationEvent: Sendable, Equatable {
    public let kind: String
    public let elementId: String?
    public let payload: String?

    public init(kind: String, elementId: String?, payload: String?) {
        self.kind = kind
        self.elementId = elementId
        self.payload = payload
    }
}

// MARK: - InMemoryUiAutomationDriver

/// Real-but-virtual UIA driver. Click / Type / Key raise `UiAutomationEvent`s
/// the host can observe. Port of C# `InMemoryUiAutomationDriver`.
public final class InMemoryUiAutomationDriver: IUiAutomationDriver, @unchecked Sendable {
    public typealias Observer = @Sendable (UiAutomationEvent) -> Void

    private let lock = NSLock()
    private var elements: [String: UiElement] = [:]
    private var observers: [Observer] = []

    public init() {}

    public var backendId: String { "in-memory" }

    /// Registers (or replaces) an element by id.
    public func register(_ element: UiElement) {
        lock.lock(); defer { lock.unlock() }
        elements[element.elementId] = element
    }

    /// Subscribes an observer to every raised event.
    public func observe(_ observer: @escaping Observer) {
        lock.lock(); defer { lock.unlock() }
        observers.append(observer)
    }

    public func snapshot() async throws -> [UiElement] {
        lock.lock(); defer { lock.unlock() }
        return Array(elements.values)
    }

    public func click(_ elementId: String) async throws {
        if elementId.trimmingCharacters(in: .whitespaces).isEmpty {
            throw UiAutomationError.elementIdRequired
        }
        let known: Bool = {
            lock.lock(); defer { lock.unlock() }
            return elements[elementId] != nil
        }()
        if !known { throw UiAutomationError.unknownElement(elementId) }
        notify(UiAutomationEvent(kind: "click", elementId: elementId, payload: nil))
    }

    public func type(_ text: String) async throws {
        // C# throws ArgumentNullException on null; Swift String is non-optional,
        // so there is nothing to reject here — proceed and raise the event.
        notify(UiAutomationEvent(kind: "type", elementId: nil, payload: text))
    }

    public func key(_ keyName: String) async throws {
        if keyName.trimmingCharacters(in: .whitespaces).isEmpty {
            throw UiAutomationError.keyNameRequired
        }
        notify(UiAutomationEvent(kind: "key", elementId: nil, payload: keyName))
    }

    /// Copies the observer list under the lock, then invokes each outside the
    /// lock, swallowing any error the observer raises (parity with the C#
    /// try/catch that logs and continues).
    private func notify(_ event: UiAutomationEvent) {
        let snapshot: [Observer] = {
            lock.lock(); defer { lock.unlock() }
            return observers
        }()
        for observer in snapshot {
            observer(event)
        }
    }
}

// MARK: - NullUiAutomationDriver

/// No-op driver. Snapshot is empty; all actions succeed silently. Port of C#
/// `NullUiAutomationDriver`.
public final class NullUiAutomationDriver: IUiAutomationDriver, @unchecked Sendable {
    public static let shared = NullUiAutomationDriver()

    public init() {}

    public var backendId: String { "null" }
    public func snapshot() async throws -> [UiElement] { [] }
    public func click(_ elementId: String) async throws {}
    public func type(_ text: String) async throws {}
    public func key(_ keyName: String) async throws {}
}

// MARK: - UiElementHelpers

/// Helpers for building / querying `UiElement`s. Port of C# static
/// `UiElementHelpers`.
public enum UiElementHelpers {
    /// True when (x, y) falls inside the element's bounds (right/bottom exclusive).
    public static func containsPoint(_ element: UiElement, x: Int, y: Int) -> Bool {
        x >= element.x && y >= element.y && x < element.x + element.width && y < element.y + element.height
    }

    /// All elements whose bounds contain (x, y).
    public static func hitTest(_ elements: [UiElement], x: Int, y: Int) -> [UiElement] {
        elements.filter { containsPoint($0, x: x, y: y) }
    }

    /// Formatted multi-line dump for debugging. Matches the C# `Dump` layout:
    /// `id "name" kind @ (x,y) WxH\n` per element.
    public static func dump(_ elements: [UiElement]) -> String {
        var s = ""
        for e in elements {
            s += "\(e.elementId) \"\(e.name)\" \(e.kind) @ (\(e.x),\(e.y)) \(e.width)x\(e.height)\n"
        }
        return s
    }
}

// MARK: - UiElement convenience

extension UiElement {
    /// Instance-style hit-test mirroring the C# extension `ContainsPoint`.
    public func contains(x: Int, y: Int) -> Bool {
        UiElementHelpers.containsPoint(self, x: x, y: y)
    }
}

File: kotlin/src/main/kotlin/com/bhengubv/circleai/windowsautomation/WindowsAutomation.kt
// WindowsAutomation.kt
//
// Kotlin port of CircleAI.WindowsAutomation — the C# reference is the EXACT
// spec. Portable UI-automation contract plus a real-but-virtual in-memory
// driver and a null driver. The platform-specific Win32-UIA backend is NOT
// portable and is intentionally excluded; hosts inject a native
// [IUiAutomationDriver] on Windows. This module ports the contract + logic only.
//
// Covers (C# file -> Kotlin type):
//   Contracts.cs                  -> UiElement, IUiAutomationDriver
//   InMemoryWindowsAutomation.cs  -> UiAutomationEvent, InMemoryUiAutomationDriver
//   NullImplementations.cs        -> NullUiAutomationDriver
//   WindowsAutomationHelpers.cs   -> UiElementHelpers (containsPoint/hitTest/dump)
//
// Fidelity notes:
//   * C# `record` -> Kotlin `data class`.
//   * C# `ValueTask` / `ValueTask<T>` -> `suspend fun` / `suspend fun (...) : T`.
//   * C# `Action<UiAutomationEvent>` observer -> `(UiAutomationEvent) -> Unit`.
//   * C# `ConcurrentDictionary` (Ordinal) -> `ConcurrentHashMap`; the observer
//     list lives behind a lock, mirroring the C# `List` + `lock`.
//   * C# extension methods (`this UiElement`) -> Kotlin extension functions.
//   * Argument validation mirrors the C# throws:
//       - blank elementId/keyName -> IllegalArgumentException
//       - unknown element on click -> IllegalStateException

package com.bhengubv.circleai.windowsautomation

import java.util.concurrent.ConcurrentHashMap

// =====================================================================
// Contracts (Contracts.cs)
// =====================================================================

/** A single element in a UI-automation snapshot. */
data class UiElement(
    val elementId: String,
    val name: String,
    val kind: String,
    val x: Int,
    val y: Int,
    val width: Int,
    val height: Int,
)

/**
 * A UI-automation backend. Hosts snap a real Win32-UIA implementation in for
 * production; portable code drives it through this contract.
 */
interface IUiAutomationDriver {
    val backendId: String
    suspend fun snapshot(): List<UiElement>
    suspend fun click(elementId: String)
    suspend fun type(text: String)
    suspend fun key(keyName: String)
}

// =====================================================================
// InMemoryWindowsAutomation (InMemoryWindowsAutomation.cs)
// =====================================================================

/** An event raised by [InMemoryUiAutomationDriver] on click / type / key. */
data class UiAutomationEvent(val kind: String, val elementId: String?, val payload: String?)

/**
 * Real-but-virtual UIA driver. Lets tests drive a virtual UI without touching
 * the desktop. Click + Type + Key raise events the host can observe.
 */
class InMemoryUiAutomationDriver : IUiAutomationDriver {
    private val elements = ConcurrentHashMap<String, UiElement>()
    private val observers = ArrayList<(UiAutomationEvent) -> Unit>()
    private val lock = Any()

    override val backendId: String get() = "in-memory"

    fun register(el: UiElement) {
        elements[el.elementId] = el
    }

    fun observe(obs: (UiAutomationEvent) -> Unit) {
        synchronized(lock) { observers.add(obs) }
    }

    override suspend fun snapshot(): List<UiElement> = elements.values.toList()

    override suspend fun click(elementId: String) {
        require(elementId.isNotBlank()) { "elementId required" }
        check(elements.containsKey(elementId)) { "Unknown element '$elementId'." }
        notify(UiAutomationEvent("click", elementId, null))
    }

    override suspend fun type(text: String) {
        notify(UiAutomationEvent("type", null, text))
    }

    override suspend fun key(keyName: String) {
        require(keyName.isNotBlank()) { "keyName required" }
        notify(UiAutomationEvent("key", null, keyName))
    }

    private fun notify(ev: UiAutomationEvent) {
        val snap: List<(UiAutomationEvent) -> Unit> = synchronized(lock) { observers.toList() }
        for (o in snap) {
            try {
                o(ev)
            } catch (ex: Exception) {
                // Mirrors the C# Debug.WriteLine: an observer must not break the driver.
                System.err.println("[CircleAI.WindowsAutomation] UI observer threw: ${ex.message}")
            }
        }
    }
}

// =====================================================================
// NullImplementations (NullImplementations.cs)
// =====================================================================

/** No-op [IUiAutomationDriver] — snapshots empty, all actions succeed silently. */
class NullUiAutomationDriver private constructor() : IUiAutomationDriver {
    override val backendId: String get() = "null"
    override suspend fun snapshot(): List<UiElement> = emptyList()
    override suspend fun click(elementId: String) {}
    override suspend fun type(text: String) {}
    override suspend fun key(keyName: String) {}

    companion object {
        val Instance = NullUiAutomationDriver()
    }
}

// =====================================================================
// WindowsAutomationHelpers (WindowsAutomationHelpers.cs)
// =====================================================================

/** Returns true when the point ([x], [y]) falls inside this element's bounds. */
fun UiElement.containsPoint(x: Int, y: Int): Boolean =
    x >= this.x && y >= this.y && x < this.x + width && y < this.y + height

/** Returns every element in [elements] whose bounds contain the point ([x], [y]). */
fun hitTest(elements: Iterable<UiElement>, x: Int, y: Int): List<UiElement> =
    elements.filter { it.containsPoint(x, y) }

/** Formats [elements] as a newline-delimited debug dump. */
fun dump(elements: Iterable<UiElement>): String {
    val sb = StringBuilder()
    for (e in elements) {
        sb.append(e.elementId).append(" \"").append(e.name).append("\" ")
            .append(e.kind).append(" @ (").append(e.x).append(",").append(e.y)
            .append(") ").append(e.width).append('x').append(e.height).append('\n')
    }
    return sb.toString()
}
